### ISSUE TITLE
(Upstream ?) Bug When Upload a Metadata File for A  Layer

### ISSUE THREAD

**Issue by ismailsunni** [kartoza/geonode#251]
While investigating on #143 I found this weird behaviour, here the steps that I took:
1. Upload a layer (with the metadata file)
2. Go to layer page, Edit Layer, and Upload Metadata
3. You will be redirect to upload metada page, select your new metadata and click upload

I expect, the new metadata will replace the old one. But, it create new layer instead, with xml file only.

It's possible that this bug is upstream bug, but I don't have any geonode with geoserver instance to test.

Have you ever experience this thing @NyakudyaA @gubuntu @lucernae

**Comment by ismailsunni** [comment]
In stagging, I even got an error:
<img width="1312" alt="screen shot 2017-06-07 at 12 34 21" src="https://user-images.githubusercontent.com/1421861/26863590-b7863aa8-4b7d-11e7-974a-036f22c61ce5.png">

**Comment by gubuntu** [comment]
The ability to upload standalone metadata arrived in 2.6 (I guess mainly to reference remote data sources).

But I would expect if you upload a new metadata document as an update to existing metadata that it would replace the existing record and remain associated with the same layer. If it's not doing that I think it's an upstream bug.

**Comment by ismailsunni** [comment]
I got different error when I tried to upload a metadata in the geoserver backend one.
<img width="1334" alt="screen shot 2017-06-08 at 14 35 54" src="https://user-images.githubusercontent.com/1421861/26917289-e6dacabe-4c57-11e7-8ee9-3129ea828d77.png">

And it doesn't create new layer instance also.

**Comment by NyakudyaA** [comment]
@ismailsunni  yes I have experienced this behavior.  but with GeoServer, it actually uploads the xml and correctly associates it with the layer. The metadata uploaded this way cannot be edited

**Comment by gubuntu** [comment]
@NyakudyaA please test again and comment or close

**Comment by NyakudyaA** [comment]
Yes this error still exists. We need to fix this.

**Comment by gubuntu** [comment]
we need to make sure this works with QGIS server backend to support #143, asap

**Comment by lucernae** [comment]
@gubuntu I read through the code. It doesn't seem upstream code properly handles this or it maybe too deep
What I tried is this:

1. Upload the same metadata from what is previously used to upload layer:

Doesn't work, the metadata document needs an identifier to properly specify which layer it belongs to in GeoNode

2. Download the metadata from Layers Details, then upload it

Doesn't work, the xml can't be parsed. Probably because of some string in the Links reference.

3. Download the metadata from Layers Details. Delete Links reference (because it will be created or the same anyway). Upload it.

I look at file_upload code, there is no reference that the new metadata info (it was successfully read), being used to update the information for the layer. I think GeoServer uses a different mechanism to update it. We might need to patch this on upstream.

**Comment by gubuntu** [comment]
This sounds like another component of GeoNode that needs to be abstracted and decoupled from Geoserver so it is back-end agnostic. 

@lucernae what do you think is the simplest thing we can do to get metadata upload and replace working, even if it is an upstream patch?

- there should not be anything Geoserver or QGIS-specific linking a metadata document (or record) to a layer
- there should be valid OWS service references in the appropriate metadata fields, auto populated and maintained by GeoNode
- there should be an identifier associating a metadata document (or record) to a GeoNode layer if indeed it describes a GeoNode layer. If it describes a remote resource then not (unless the remote resource gets registered by GeoNode as a virtual layer?)

**Comment by gubuntu** [comment]
@lucernae hose does https://github.com/kartoza/geosafe/pull/378 affect this?
